modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/consensus/ConsensusStateUpdater.scala

```diff
@@ -0,0 +1,182 @@
+    ): F[Option[(Boolean, ConsensusState[Key, Artifact])]] =
+      maybeState.flatTraverse { state =>
+        state.status match {
+          case Facilitated() =>
+            val maybeBound = state.facilitators
+              .traverse(resources.peerDeclarations.get)
+              .flatMap(_.foldMap(_.upperBound))
+
+            maybeBound.traverse { bound =>
+              for {
+                peerEvents <- consensusStorage.pullEvents(bound)
+                events = peerEvents.toList.flatMap(_._2).map(_._2).toSet
+                (artifact, returnedEvents) <- consensusFns
+                  .createProposalArtifact(state.lastKeyAndArtifact, events)
+                returnedPeerEvents = peerEvents.map {
+                  case (peerId, events) =>
+                    (peerId, events.filter { case (_, event) => returnedEvents.contains(event) })
+                }.filter { case (_, events) => events.nonEmpty }
+                _ <- consensusStorage.addEvents(returnedPeerEvents)
+                hash <- artifact.hashF
+                newState = state.copy(status = ProposalMade[Artifact](hash, artifact))
+                result <- setter(newState.some)
+                _ <- gossip.spread(ConsensusProposal(key, hash))
+              } yield (result, newState)
+            }
+          case ProposalMade(proposalHash, proposalArtifact) =>
+            val maybeMajority = state.facilitators
+              .traverse(resources.peerDeclarations.get)
+              .flatMap(_.traverse(_.proposal))
+              .flatMap(pickMajority)
+
+            maybeMajority.traverse { majorityHash =>
+              for {
+                _ <- Applicative[F].unit
+                newState = state.copy(status = MajoritySelected[Artifact](majorityHash))
+                result <- setter(newState.some)
+                signature <- Signature.fromHash(keyPair.getPrivate, majorityHash)
+                _ <- gossip.spread(MajoritySignature(key, signature))
+                _ <- if (majorityHash == proposalHash)
+                  gossip.spreadCommon(ConsensusArtifact(key, proposalArtifact))
+                else
+                  Applicative[F].unit
+              } yield (result, newState)
+
+            }
+          case MajoritySelected(majorityHash) =>
+            val maybeAllSignatures =
+              state.facilitators.sorted.traverse { peerId =>
+                resources.peerDeclarations
+                  .get(peerId)
+                  .flatMap(peerDeclaration => peerDeclaration.signature.map(signature => (peerId, signature)))
+              }.map(_.map { case (id, signature) => SignatureProof(PeerId._Id.get(id), signature) })
+
+            val maybeSignedArtifact = for {
+              allSignatures <- maybeAllSignatures
+              allSignaturesNel <- NonEmptyList.fromList(allSignatures)
+              majorityArtifact <- resources.artifacts.get(majorityHash)
+            } yield Signed(majorityArtifact, allSignaturesNel)
```

Comment:
you can consider such a syntax as well:

```scala
(maybeAllSignatures.flatMap(NonEmptyList.fromList), resources.artifacts.get(majorityHash))
  .mapN(Signed(_, _)
```

Reply:
Looks shorter, but Idk about the readability. I think that `for` in that case actually is more readable.